python/paddle/amp/auto_cast.py

```diff
@@ -25,10 +36,41 @@
     in_pir_mode,
 )
 from paddle.base.wrapped_decorator import signature_safe_contextmanager
+from paddle.distributed.fleet.meta_optimizers.dygraph_optimizer.dygraph_sharding_optimizer import (
+    DygraphShardingOptimizer,
+    DygraphShardingOptimizerV2,
+)
 from paddle.static.amp.decorator import OptimizerWithMixedPrecision
 
 from .amp_lists import black_list, white_list
 
+if TYPE_CHECKING:
+    from typing import Generator
+
+    from paddle import Tensor, dtype
+    from paddle.nn.layer.layers import Layer
+
+    _CustomList: TypeAlias = Union[list, tuple, set]
+    _LayerList: TypeAlias = Union[
+        paddle.nn.BatchNorm,
+        paddle.nn.BatchNorm1D,
+        paddle.nn.BatchNorm2D,
+        paddle.nn.BatchNorm3D,
+        paddle.nn.SyncBatchNorm,
+        paddle.nn.LayerNorm,
+        paddle.nn.InstanceNorm1D,
+        paddle.nn.InstanceNorm2D,
+        paddle.nn.InstanceNorm3D,
+    ]
+    _OptimizerBase: TypeAlias = Union[
+        paddle.optimizer.Optimizer,
+        DygraphShardingOptimizer,
+        DygraphShardingOptimizerV2,
+    ]
```

Comment:
```suggestion
class _OptimizerLike(Protocol):
    def minimize(
        self,
        loss: Tensor,
        startup_program: Program,
        parameters: list[Tensor],
        no_grad_set: set[Tensor],
    ) -> tuple[list[Operator], list[tuple[Tensor, Tensor]]]: ...
    def step(self) -> None: ...
    def set_state_dict(self, state_dict: dict[str, Tensor]) -> None: ...
    def clear_grad(self, set_to_zero: bool) -> None: ...
```

这里定义一个 Protocol ，然后后面都用这个就行 ～

最好的办法是 Optimizer 和这里的 DygraphShardingOptimizer 都有一个公共父类，既然没有，我们定义一个 Protocol 暂时解决一下吧 ～

Reply:
@enkilee `_OptimizerLike` 改了一下，减少限制，可以试一下 ～

Reply:
收到